Import data from google spreadsheet

Prakriti-nith opened the pull request:
#87 
This PR adds a feature to retrieve data from google spreadsheets.

Prakriti-nith commented:
Once the code is reviewed, I will add the tests for the related methods.

Shekharrajak commented:
Writing testcases before writing lines of code is good practice. So testcases must be present for each line of code (test coverage 100%).

Shekharrajak commented on spec/adapters/googlecharts_spec.rb:
```diff
@@ -30,6 +30,14 @@
       ['Lg. Magellanic Cloud', 50000, 120.9],
       ['Bootes I', 60000, 1223.1]
     ]
+    @query_string = 'SELECT A, H, O, Q, R, U LIMIT 5 OFFSET 8'
```
Please write  examples in `dummy_iruby` and share the link ( [like](http://nbviewer.jupyter.org/github/sciruby/daru-view/tree/master/spec/dummy_iruby/)).

Shekharrajak replied:
I  am expecting notebook link like [this](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/import_spreadsheet/spec/dummy_iruby/Google%20Chart%20%7C%20Data%20from%20google%20spreadsheet.ipynb) .

Prakriti-nith replied:
Here is the [link](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/d2bb96be63aaeac2f79cebbac748162c89f85daa/spec/dummy_iruby/Google%20Chart%20|%20Data%20from%20google%20spreadsheet.ipynb) to view the examples.
I am getting the outputs for all the jupyter notebook cells locally but I can't understand why they are not visible online.

Shekharrajak replied:
Did you try in incognito mode ?

Prakriti-nith replied:
Yes, I tried the examples in incognito mode. Running each cell individually is showing the correct output but after I save the file and open it again, then only the last output is visible.

Shekharrajak commented on lib/daru/view/adapters/googlecharts.rb:
```diff
@@ -46,17 +40,30 @@ def init_table(data=[], options={})
 
         def generate_body(plot)
-          plot.to_html
+          plot.to_html(@data)
```
Any  alternate way, you can think of ?

Shekharrajak commented on lib/daru/view/adapters/googlecharts/display.rb:
```diff
@@ -55,5 +71,41 @@ class DataTable
+    # Parameters:
+    #  *data              [Required] The URL of the spreadsheet in a specified format. Query string can be appended
+    #                               to retrieve the data accordingly.
+    #  *div_id            [Required] The ID of the DIV element that the Google Chart should be rendered in.
```
line size ?

Shekharrajak commented on lib/daru/view/adapters/googlecharts.rb:
```diff
@@ -47,10 +41,23 @@ def init_table(data=[], options={})
+            data.table
+          elsif data.is_a?(GoogleVisualr::DataTable)
+            data
+          elsif data.is_a?(String)
```
Check whether it is a valid URL or not. If not rescue the Exception + document this enhancement.

Prakriti-nith replied:
Where should I write the documentation?

Shekharrajak replied:
[here](https://github.com/SciRuby/daru-view/pull/88/files#diff-cd0c826056999c96e0390d69150cc07bR18)

Prakriti-nith commented:
I have resolved the problem of only last chart visibility. We need to have the separate `handle_query_response` function for each query ([refer](https://stackoverflow.com/questions/21859917/how-to-get-two-different-google-charts-on-one-page-from-two-separate-spreadsheet)).
For drawing the chart when data is imported from Google spreadsheet, we are not using google_visualr as we can't feed the data into the table. The query automatically generates the table when `var data_table = response.getDataTable();` js code is run. 
I have also improved the documentation according to the yard and added some more tests.
Updated examples: [link](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/7e49295e1a6c1d098be0a07f9d3ac1332d475879/spec/dummy_iruby/Google%20Chart%20|%20Data%20from%20google%20spreadsheet.ipynb)

Prakriti-nith commented:
@Shekharrajak ping!

Shekharrajak commented on lib/daru/view/adapters/googlecharts.rb:
```diff
@@ -97,7 +121,7 @@ def extract_chart_type(options)
           chart_type = 'SteppedArea' if chart_type == 'Steppedarea'
           chart_type = 'TreeMap' if chart_type == 'Treemap'
-          direct_name = %w[Map Histogram TreeMap Timeline Gauge]
+          direct_name = %w[Map Histogram TreeMap Timeline Gauge Table]
```
It is only for `chart`;  not for table. Where do you need this ?

Shekharrajak commented on lib/daru/view/adapters/googlecharts.rb:
```diff
@@ -14,24 +14,32 @@ module GooglechartsAdapter
-            else
-              add_data_in_table(data)
-            end
+          @table = get_table(data) unless data.is_a?(String)
```
What will happen if string is not a URL?

Shekharrajak replied:
What will be `plot.table` when we have URL?

Prakriti-nith replied:
plot.table will contain the empty table as the DataTable is generated in query response [here](https://github.com/SciRuby/daru-view/pull/88/files#diff-873104e004ae7bc85b74e657779a5804R50) and we can not retrieve the data from google spreadsheet.

Prakriti-nith replied:
We might be able to retrieve the data from the URL provided. It is because the URL provides the access to JSON data which can then be stored in a Daru::DataFrame.

Shekharrajak replied:
We can get the table , [as here](https://github.com/SciRuby/daru-view/pull/88/files#diff-cd0c826056999c96e0390d69150cc07bR107).

Shekharrajak commented on spec/adapters/googlecharts/base_chart_spec.rb:
```diff
@@ -0,0 +1,54 @@
+describe GoogleVisualr::BaseChart do
+  before { Daru::View.plotting_library = :googlecharts }
+  before(:each) do
+    @query_string = 'SELECT A, H, O, Q, R, U LIMIT 5 OFFSET 8'
```
We must use `let` instead of `@`. 
Please follow [better specs](http://www.betterspecs.org/).

Shekharrajak commented on lib/daru/view/adapters/googlecharts/display.rb:
```diff
@@ -18,23 +19,59 @@ module Display
+    # @param dom [String] The ID of the DIV element that the Google
+    #   Chart should be rendered in
+    # @return [String] js code to render the chart
+    def get_html(dom)
```
I think, it is same as `to_js` method.

Prakriti-nith replied:
Actually, I was going to ask you about this. I wanted to know why are we generating the googlecharts' code with script tag (by calling `to_js` method) and without script tag (by calling `load_js` and `draw_js`) separately in display.rb. We have already defined script tags in the template so we'll definitely be generating the script without the script tags (using load_js and draw_js)?
`to_js` method is never actually called.

Shekharrajak commented on lib/daru/view/adapters/googlecharts.rb:
```diff
@@ -14,24 +14,32 @@ module GooglechartsAdapter
+        # @param options [Hash] Various options provided by the user to
+        #   incorporate in google charts
+        # @return [GoogleVisualr::Interactive] Returns the chart object based
+        #   on the chart_type
```
Thanks for writing this. Please write few examples as well for `init` and `init_table`.

Prakriti-nith commented:
@Shekharrajak  please review the changes.

Prakriti-nith commented on lib/daru/view/adapters/googlecharts.rb:
```diff
@@ -14,24 +15,100 @@ module GooglechartsAdapter
+        #          '_7zOFDcg8Wm8Xv29_8PWuuW15qmAE/gviz/tq?gid=0&headers=1&tq='
+        #   query = 'SELECT A, H, O, Q, R, U LIMIT 5 OFFSET 8'
+        #   data << query
+        #   chart = Daru::View::Table.new(data)
```
I have already added example for this. I have further added an example for the previous one. [Link](http://nbviewer.jupyter.org/github/Prakriti-nith/daru-view/blob/d65555510d7de87a19b93b5ca7fc078280140288/spec/dummy_iruby/Google%20Chart%20|%20Data%20from%20google%20spreadsheet.ipynb).

Shekharrajak commented:
Thanks @Prakriti-nith , for writing in [wiki page](https://github.com/SciRuby/daru-view/wiki/GoogleCharts-Features)
